simba/rotation.py

```diff
@@ -27,52 +27,80 @@ def add_trip(self, trip):
-            assert self.schedule.vehicle_types.get(
-                self.vehicle_type, {}).get(charging_type) is not None, (
-                f"The required vehicle type {self.vehicle_type}({charging_type}) "
-                "is not given in the vehicle_types.json file.")
-            if self.charging_type is None:
-                # set CT for whole rotation
-                self.set_charging_type(charging_type)
-            elif self.charging_type == charging_type:
-                # same CT as other trips: just add trip consumption
-                self.consumption += new_trip.calculate_consumption()
-            else:
-                # different CT than rotation: error
-                raise Exception(f"Two trips of rotation {self.id} have distinct charging types")
+                self.arrival_time = trip.arrival_time
+                self.arrival_name = trip.arrival_name
+
+        if self.vehicle_type is None:
+            self.vehicle_type = trip.vehicle_type
+            for prior_trip in self.trips:
+                prior_trip.vehicle_type = trip.vehicle_type
+        else:
+            assert self.vehicle_type == trip.vehicle_type, (
+                f"Two trips of rotation {self.id} have distinct vehicle types")
+
+        if self.charging_type is not None and self.charging_type != trip.charging_type:
+            raise ValueError(f"Two trips of rotation {self.id} have distinct charging types")
+
+        # (re)calculate consumption
+        # might change if charging_type changes
+        # therefore, add new trip only after charging_type is set
+        trip.rotation = self
+        if trip.charging_type is None:
+            trip.charging_type = self.charging_type
+        if trip.charging_type in ["oppb", "depb"]:
+            self.set_charging_type(trip.charging_type)
+            self.consumption += trip.calculate_consumption()
+
+        self.trips.append(trip)
+
+    def add_trip_from_dict(self, trip_dict):
+        """ Create a trip object from dictionary and append to rotations trip set.
+
+        See :py:class:`~simba.trip.Trip` for arguments and defaults.
+
+        :param trip_dict: vars of Trip
+        :type trip_dict: dict
+        """
+        trip = Trip(
+            departure_time=trip_dict["departure_time"],
+            departure_name=trip_dict["departure_name"],
+            arrival_time=trip_dict["arrival_time"],
+            arrival_name=trip_dict["arrival_name"],
+            distance=trip_dict["distance"],
+            line=trip_dict.get("line"),
+            temperature=trip_dict.get("temperature"),
+            level_of_loading=trip_dict.get("level_of_loading"),
+            mean_speed=trip_dict.get("mean_speed"),
+            height_diff=trip_dict.get("height_diff"),
+            station_data=trip_dict.get("station_data"),
+            charging_type=trip_dict.get("charging_type"),
+            vehicle_type=trip_dict.get("vehicle_type"),
+        )
+        self.add_trip(trip)
```

Comment:
right now these lines seem equivalent to 
del trip_dict["rotation_id"]
self.add_trip(Trip(**trip_dict))

I think the function would be well suited to implement the type checking and casting and default value implementation, e.g. distance = float(trip_dict["distance"]), etc. This would allow for proper type declaration in Trip()

Reply:
moved casts from Trip init to Rotation.add_trip_from_dict